### PR TITLE
TNO-2083 Limit users to signing in on one device

### PR DESCRIPTION
**Pull request by Fosol**
User accounts can now be configured to limit the number of concurrent connections.

## Summary

- Added way to control whether a user can only log into one device
- Updated Editor app
- Updated Subscriber app
- Added DB migration 1.0.105
- Published tno-core:0.1.1

## Auto Detection

image

### CODE REVIEW

File: api/net/Controllers/AuthController.cs [hunk 1]
```diff
@@ -59,12 +54,28 @@ public AuthController(ICssHelper cssHelper, IUserService userService, IOptions<J
     /// <returns></returns>
     [HttpPost("userinfo")]
     [Produces(MediaTypeNames.Application.Json)]
-    [ProducesResponseType(typeof(PrincipalModel), 200)]
+    [ProducesResponseType(typeof(PrincipalModel), (int)HttpStatusCode.OK)]
     [SwaggerOperation(Tags = new[] { "Auth" })]
-    public async Task<IActionResult> UserInfoAsync()
+    public async Task<IActionResult> UserInfoAsync(LocationModel? location)
```
Comment (Fosol) [inline api/net/Controllers/AuthController.cs:59]: Optionally we can supply the user location.  
Later I'll extract the user IP from the request to the API to ensure it isn't blocked.

File: api/net/Keycloak/CssHelper.cs [hunk 2]
```diff
@@ -175,21 +180,182 @@ private async Task AddOrUpdateUserAsync(string username, Entities.User? user, TN
...
+    }
+
+    /// <summary>
+    /// If a location is provided it will add it to the user preferences property.
+    /// Determine if the user has logged into too many devices at the same time.
+    /// </summary>
+    /// <param name="user"></param>
+    /// <param name="location"></param>
+    /// <returns></returns>
+    private static Models.Auth.AccountAuthState AuthorizeLocation(Entities.User user, Models.Auth.LocationModel? location = null)
```
Comment (Fosol) [inline api/net/Keycloak/CssHelper.cs:270]: This is the magic function

File: app/subscriber/src/features/login/UserInfo.tsx [hunk 3]
```diff
@@ -14,19 +22,56 @@ export interface IUserInfoProps {
...
       appStore.storeUserInfo();
       setInit(true);
     }
-  }, [app, appStore, keycloak.authenticated, init]);
+  }, [app, appStore, keycloak.authenticated, init, toggle]);
 
-  return <>{children}</>;
+  hub.useHubEffect(MessageTargetName.Logout, (message: any) => {
+    if (!isShowing) toggle();
+    else keycloak.instance.logout();
```
Comment (Fosol) [inline app/subscriber/src/features/login/UserInfo.tsx:47]: Logout when receiving the logout message

File: app/subscriber/src/features/settings/MyMinisterSettings.tsx [hunk 4]
```diff
@@ -17,15 +17,18 @@ export const MyMinisterSettings: React.FC = () => {
...
     .sort((a, b) => a.sortOrder - b.sortOrder || a.name.localeCompare(b.name));
 
-  const handleSubmit = async (values: number[]) => {
+  const handleSubmit = async (values: number[], userInfo: IUserInfoModel) => {
     try {
-      const user = {
+      var user = {
         ...userInfo,
-        preferences: { ...userInfo?.preferences, myMinisters: values },
+        preferences: { ...userInfo.preferences, myMinisters: values },
```
Comment (Fosol) [inline app/subscriber/src/features/settings/MyMinisterSettings.tsx:24]: This is why javascript is horrible.  Don't do this kind of fake merge.  We need a proper endpoint that updates the user preferences.

File: app/subscriber/src/features/settings/MyMinisterSettings.tsx [hunk 5]
```diff
@@ -37,25 +40,32 @@ export const MyMinisterSettings: React.FC = () => {
...
         const isActive = activeMinisters.find((element) => element.id === m);
         if (isActive) activeSelectedMinisters.push(m);
         else inactiveSelectedMinisters.push(m);
       });
       if (inactiveSelectedMinisters.length !== 0) {
-        const user = {
+        var user = {
           ...userInfo,
-          preferences: { ...userInfo?.preferences, myMinisters: activeSelectedMinisters },
+          preferences: { ...userInfo.preferences, myMinisters: activeSelectedMinisters },
```
Comment (Fosol) [inline app/subscriber/src/features/settings/MyMinisterSettings.tsx:54]: Same problem here.

File: app/subscriber/src/store/hooks/app/useApp.ts [hunk 6]
```diff
@@ -82,7 +86,25 @@ export const useApp = (): [IAppState, IAppController] => {
...
+        try {
+          // Generate a unique key for this user and store in local storage.
+          var key = localStorage.getItem('device-key');
+          if (!key) {
+            key = crypto.randomUUID();
+            localStorage.setItem('device-key', key);
+          }
+          const locationResponse = await dispatch(
+            'get-location',
+            () => axios.get('https://geolocation-db.com/json/'),
```
Comment (Fosol) [inline app/subscriber/src/store/hooks/app/useApp.ts:99]: I don't think this is a great solution, but it's a neat one.

File: api/net/Keycloak/CssHelper.cs (hunk 2 above)
Comment (jdtoombs) [reply api/net/Keycloak/CssHelper.cs:270]: cool!

**Review by jdtoombs — APPROVED** [review]

File: app/subscriber/src/store/hooks/app/useApp.ts (hunk 6 above)
Comment (jdtoombs) [reply app/subscriber/src/store/hooks/app/useApp.ts:99]: Didn't even know that existed - 'tis neat!

**Pull request merged**